dlio_benchmark/checkpointing/base_checkpointing.py

```diff
@@ -21,53 +21,207 @@
+        if l <= 0:
+            return 0
+        embedding = voc*h
+        input_norm = h
+        qkv = 3*h*h
+        dense = h*h
+        layer_norm = h
+        mlp_h_to_4h = ffn*2*h # the factor of 2 is because of gated linear unit                                                                           
+        mlp_4h_to_h = ffn*h
+        weight = h
+        lm_head = embedding
+        return embedding  + (input_norm + qkv + dense + layer_norm + mlp_h_to_4h + mlp_4h_to_h)*l + weight + lm_head
+
+    def get_layer_parameters(self, layer_index):
+        if len(self.args.layer_parameters) > 0:
+            self.layer_parameters_predefined = True
+            return self.args.layer_parameters
+        else:
+            if self.args.num_layers <= 0:
+                return []
+            if self.args.zero_stage < 3:
+                sharding_factor = 1
+            else:
+                sharding_factor = self.dp
+            h, l, ffn, voc = self.args.hidden_size, self.args.num_layers, self.args.ffn_hidden_size, self.args.vocab_size
+            if layer_index == 0 or layer_index == l + 1:
+                return [h * voc // self.tp // sharding_factor] # embedding or lm_head
+            elif layer_index == l + 2:
+                return [h//sharding_factor]
+            else:
+                return [ h // sharding_factor, # input_norm, 
+                        h*h*3//self.tp//sharding_factor, # self_attn
+                        h*h//self.tp//sharding_factor, # dense
+                        h//sharding_factor, # layer_norm
+                        h*2*ffn//self.tp//sharding_factor, # ffn_h_to_4h, 2 is from gated linear unit
+                        h*ffn//self.tp//sharding_factor, # ffn_4h_to_h
+                ]
+    def get_layer_state(self, layer_index):
+        layer_parameters = self.get_layer_parameters(layer_index)
+        layer_state = dict()
+        size = 0.0
+        for index, state in enumerate(layer_parameters):
+            if state > 0:
+                layer_state[str(index)] = self.get_tensor(state, self.args.checkpoint_model_datatype)
+                size += state*get_datatype_size(self.args.checkpoint_model_datatype)
+        return layer_state, size
+
+    def get_optimization_groups(self):
+        h, l, ffn, voc = self.args.hidden_size, self.args.num_layers, self.args.ffn_hidden_size, self.args.vocab_size
+        if len(self.args.optimization_groups) > 0:
+            self.optimization_groups_predefined = True
+            return self.args.optimization_groups
+        else:
+            if self.args.num_layers <= 0:
+                return []
+            if self.args.zero_stage > 0:
+                # zero stage 1, 2, 3
+                num_p = self.get_num_parameters() // self.args.comm_size
+            else:
+                # if zero is not used. Only the first data parallel instance will save the optimizer states
+                num_p = self.get_num_parameters() // self.mp
```

Comment:
In this code, why zero_stage=0 can't determine the first data parallel instance will save the optimizer states. We dont need -1 right? Or does zero_stage has more meaning?